rfcs/text/0010_service_status.md

```diff
@@ -0,0 +1,304 @@
+
+_Disabled_ plugins, that is plugins that are explicitly disabled in Kibana's configuration, do not have any status. They are not present in any status APIs and are **not** considered `unavailable`. Disabled plugins are excluded from the status inheritance calculation, even if a plugin has a optional dependency on a disabled plugin. In summary, if a plugin has an optional dependency on a disabled plugin, the plugin will not be considered `degraded` just because that optional dependency is disabled.
+
+### HTTP responses
+
+As specified in the [_Levels section_](#levels), a service's HTTP endpoints will respond with `503 Unavailable` responses in some status levels.
+
+In both the `fatal` and `unavailable` levels, all of a service's endpoints will return 503s. However, in the `degraded` level, it is up to service authors to decide which endpoints should return a 503. This may be implemented directly in the route handler logic or by using any of the [utilities provided](#status-utilities).
+
+## Status Utilities
+
+Though many plugins should be able to rely on the default status inheritance and associated behaviors, there are common patterns and overrides that some plugins will need. The status service should provide some utilities for these common patterns out-of-the-box.
+
+```ts
+/**
+ * Extension of the main Status API
+ */
+interface StatusSetup {
+  /**
+   * Helpers for expressing status in HTTP routes.
+   */
+  http: {
+    /**
+     * High-order route handler function for wrapping routes with 503 logic based
+     * on a predicate.
+     * 
+     * @remarks
+     * When a 503 is returned, it also includes detailed information from the service's
+     * current `ServiceStatus` including `meta` information.
+     * 
+     * @example
+     * ```ts
+     * router.get(
+     *   { path: '/my-api' }
+     *   unavailableWhen(
+     *     ServiceStatusLevel.degraded,
+     *     async (context, req, res) => {
+     *       return res.ok({ body: 'done' });
+     *     }
+     *   )
+     * )
+     * ```
+     * 
+     * @param predicate When a level is specified, if the plugin's current status
+     *                  level is >= to the severity of the  specified level, route
+     *                  returns a 503. When a function is specified, if that
+     *                  function returns `true`, a 503 is returned.
+     * @param handler The route handler to execute when a 503 is not returned.
+     */
+    unavailableWhen<P, Q, B>(
+      predicate: ServiceStatusLevel |
+        (core: CoreStatus, plugins: Record<string, ServiceStatus>) => boolean,
+      handler: RouteHandler<P, Q, B>
+    ): RouteHandler<P, Q, B>;
+  }
+}
+```
+
+## Additional Examples
+
+### Combine inherited status with check against external dependency
+```ts
+const getExternalDepHealth = async () => {
+  const resp = await window.fetch('https://myexternaldep.com/_healthz');
```

Comment:
shouldn't we use http.core.fetch in this example? we discourage `window.fetch` usage https://github.com/elastic/kibana/issues/57151

Reply:
This is to an outside service, which `core.http.fetch` does not currently support.

Reply:
> which core.http.fetch does not currently support.

![shocked](https://user-images.githubusercontent.com/3198181/76523542-e20b7700-6468-11ea-9157-359c63b7f30c.jpg)